test_data/pil/expand_fixed.pil

```diff
@@ -0,0 +1,172 @@
+    let<T> sub_array: T[], int, int -> T[] = |arr, start, l| std::array::new(l, |i| arr[start + i]);
+
+    /// Evaluates to the array [f(arr[0]), f(arr[1]), ..., f(arr[len(arr) - 1])].
+    let<T1, T2> map: T1[], (T1 -> T2) -> T2[] = |arr, f| new(len(arr), |i| f(arr[i]));
+
+    /// Evaluates to the array [f(0, arr[0]), f(1, arr[1]), ..., f(len(arr) - 1, arr[len(arr) - 1])].
+    let<T1, T2> map_enumerated: T1[], (int, T1 -> T2) -> T2[] = |arr, f| new(len(arr), |i| f(i, arr[i]));
+
+    /// Computes folder(...folder(folder(initial, arr[0]), arr[1]) ..., arr[len(arr) - 1])
+    let<T1, T2> fold: T1[], T2, (T2, T1 -> T2) -> T2 = |arr, initial, folder| std::utils::fold(len(arr), |i| arr[i], initial, folder);
+
+    /// Returns the sum of the array elements.
+    // TODO: Should make use of the Default or Zero trait instead of FromLiteral (then we can also
+    // use this function to flatten an array of arrays.
+    let<T: Add + FromLiteral> sum: T[] -> T = |arr| fold(arr, 0, |a, b| a + b);
+
+    /// Returns the product of the array elements.
+    let<T: Mul + FromLiteral> product: T[] -> T = |arr| fold(arr, 1, |a, b| a * b);
+
+    /// Zips two arrays
+    /// TODO: Assert that lengths are equal when expressions are supported.
+    let<T1, T2, T3> zip: T1[], T2[], (T1, T2 -> T3) -> T3[] = |array1, array2, fn| new(len(array1), |i| fn(array1[i], array2[i]));
+
+namespace Main(8);
+
+    // A ArrayTerm in an `ArrayExpression`
+    enum ArrayTerm {
+        Repeat(int[]),
+        Once(int[])
+    }
+
+    // returns the total size of the repeated array in this `ArrayExpression`
+    let solve: ArrayTerm[] -> int = |terms| {
+        let (_, total_size_of_non_repeated) = std::array::fold(terms, (std::prelude::false, 0), |(found_repeated, len), ArrayTerm| {
+            match ArrayTerm {
+                // we can have `[]*` as many times as we want
+                ArrayTerm::Repeat([]) => (found_repeated, len),
+                ArrayTerm::Repeat(a) => {
+                    if found_repeated {
+                        std::check::panic("unsolvable because two terms are repeated");
+                        (found_repeated, 42)
+                    } else {
+                        (std::prelude::true, len)
+                    }
+                },
+                ArrayTerm::Once(a) => (found_repeated, len + std::array::len(a))
+            }
+        });
+        if total_size_of_non_repeated > std::prover::degree() {
+            std::check::panic("non repeated array terms do not fit in degree");
+            42
+        } else {
+            std::prover::degree() - total_size_of_non_repeated
+        }
+    };
+
+    /// repeat `repeated` until we reach a length of `total_len`, truncating
+    let expand_to_len = |repeated, total_len|
+        std::array::new(total_len, |j| repeated[j % std::array::len(repeated)]);
+
+    // expand an `ArrayExpression` into a single complete array
+    let expand: ArrayTerm[] -> int[] = |terms| {
```

Comment:
I think this should return an `int -> int` function instead. Explicitly creating the big arrays in the pil evaluator should be quite expensive.